### PR TITLE
Restoring files from storage to buffer

### PR DESCRIPTION
**Pull request by iisuslik43**
Before doing smth in methods like `createPatch`, `getVersionList`, `forcedCommit` and `changeFileId`, I m restoring file to buffer from GCS if buffer doesn't contain such file

It's too early to review this code, it needs more tests

### CODE REVIEW

File: src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt [hunk 1]
```diff
@@ -481,6 +461,42 @@ class CosmasGoogleCloudService(private val freeBucketName: String,
         responseObserver.onCompleted()
     }
 
+    fun restoreFileFromStorage(fileId: String, projectInfo: ProjectInfo,
```
Comment (dbarashev) [inline src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt]: could be private

File: src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt [hunk 2]
```diff
@@ -481,6 +461,42 @@ class CosmasGoogleCloudService(private val freeBucketName: String,
...
+                               project: ConcurrentMap<String, CosmasProto.FileVersion>): FileVersion {
+        synchronized(project) {
+            // Getting last version from storage or default instance if it doesn't exist
+            val latestVersionBlob: Blob? = this.storage.get(getBlobId(fileId, projectInfo))
+            val latestVersion = if (latestVersionBlob != null) {
+                CosmasProto.FileVersion.parseFrom(latestVersionBlob.getContent())
+            } else {
+                val bufferVersion = FileVersion.getDefaultInstance()
+                project[fileId] = bufferVersion
```
Comment (dbarashev) [inline src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt]: Can we avoid modifying project here? It seems that the returned result will be written to project map anyway in the calling method, while keeping this function read-only will make code more simple and robust.

File: src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt [hunk 3]
```diff
@@ -481,6 +461,42 @@ class CosmasGoogleCloudService(private val freeBucketName: String,
         responseObserver.onCompleted()
     }
 
+    fun restoreFileFromStorage(fileId: String, projectInfo: ProjectInfo,
+                               project: ConcurrentMap<String, CosmasProto.FileVersion>): FileVersion {
+        synchronized(project) {
+            // Getting last version from storage or default instance if it doesn't exist
+            val latestVersionBlob: Blob? = this.storage.get(getBlobId(fileId, projectInfo))
+            val latestVersion = if (latestVersionBlob != null) {
```
Comment (dbarashev) [inline src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt]: This combination of `if` with `return` looks uncomfortable. I suggest appending null-case to the previous line with Elvis-operator:

```
            val latestVersionBlob: Blob = this.storage.get(getBlobId(fileId, projectInfo)) ?: return FileVersion.getDefaultInstance() 
            val latestVersion = CosmasProto.FileVersion.parseFrom(latestVersionBlob.getContent())
```

File: src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt (hunk 1 above)
Comment (iisuslik43) [reply src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt]: done

File: src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt (hunk 2 above)
Comment (iisuslik43) [reply src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt]: ok, done

File: src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt (hunk 3 above)
Comment (iisuslik43) [reply src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt]: ok

File: src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt [hunk 4]
```diff
@@ -264,9 +263,13 @@ class CosmasGoogleCloudService(private val freeBucketName: String,
             return
         }
         if (versionList.isEmpty()) {
-            val errorStatus = Status.NOT_FOUND.withDescription(
-                    "There is no file in storage with file id ${request.fileId}")
-            LOG.error(errorStatus.description)
+            val description = if (request.startGeneration != -1L) {
+                "There is no file=${request.fileId} in storage"
+            } else {
+                "There is no versions of file=${request.fileId} before generation=${request.startGeneration}"
```
Comment (dbarashev) [inline src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt:269]: will it always print "before generation=-1" ?

File: src/test/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudServiceTest.kt [hunk 5]
```diff
@@ -116,6 +117,39 @@ class CosmasGoogleCloudServiceTest {
         assertEquals("kek lol", file)
     }
 
+    @Test
+    fun restoreToBufferCreatePatch() {
+        val fakeStorage: Storage = mock(Storage::class.java)
+        val blob1 = getMockedBlob("ver1", 0)
+        val blob2 = getMockedBlob("ver2", 1)
+        Mockito.`when`(fakeStorage.get(eq(service.getBlobId(FILE_ID, projectInfo(), null))))
+                .thenReturn(blob1).thenReturn(blob2)
```
Comment (dbarashev) [inline src/test/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudServiceTest.kt]: I suggest moving the second `thenReturn` lower after call to `addPatchToService` to make it clear that we return blob2 only after that

File: src/test/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudServiceTest.kt [hunk 6]
```diff
@@ -116,6 +117,39 @@ class CosmasGoogleCloudServiceTest {
...
+
+    @Test
+    fun restoreToBufferVersionList() {
+        val fakeStorage: Storage = mock(Storage::class.java)
+        val blob1 = getMockedBlob("ver1", 0)
+        val blob2 = getMockedBlob("ver2", 1)
+        Mockito.`when`(fakeStorage.get(eq(service.getBlobId(FILE_ID, projectInfo(), null))))
+                .thenReturn(blob1).thenReturn(blob2)
+        this.service = CosmasGoogleCloudService(this.BUCKET_NAME, fakeStorage, getMockedClock())
+        assertEquals(listOf(0L), getVersionsList(FILE_ID, PROJECT_ID))
```
Comment (dbarashev) [inline src/test/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudServiceTest.kt:149]: Why there is only 1 value in the list, despite that we seem to be operating with two blobs above?

File: src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt (hunk 4 above)
Comment (iisuslik43) [reply src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt:269]: it should be `request.startGeneration == -1L`, fixed

File: src/test/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudServiceTest.kt (hunk 5 above)
Comment (iisuslik43) [reply src/test/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudServiceTest.kt]: ok

File: src/test/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudServiceTest.kt (hunk 6 above)
Comment (iisuslik43) [reply src/test/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudServiceTest.kt:149]: It was pretty strange test, fixed

**Review by dbarashev — APPROVED** [review]
LGTM

**Pull request merged**